core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingService.java

```diff
@@ -0,0 +1,92 @@
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.throttling;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.DefaultHttpResponse;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.SimpleDecoratingService;
+
+/**
+ * Decorates an HTTP {@link Service} to throttle incoming requests.
+ * TODO: provides ThrottlingRpcService and ThrottlingHttpService
+ */
+public class ThrottlingService extends SimpleDecoratingService<HttpRequest, HttpResponse> {
+
+    /**
+     * Creates a new decorator using the specified {@link ThrottlingStrategy} instance.
+     *
+     * @param strategy The {@link ThrottlingStrategy} instance to be used
+     */
+    public static Function<Service<HttpRequest, HttpResponse>, ThrottlingService>
+    newDecorator(ThrottlingStrategy strategy) {
+        requireNonNull(strategy, "strategy");
+        return delegate -> new ThrottlingService(delegate, strategy);
+    }
+
+    private final ThrottlingStrategy strategy;
+
+    /**
+     * Creates a new instance that decorates the specified {@link Service}.
+     */
+    protected ThrottlingService(Service<HttpRequest, HttpResponse> delegate, ThrottlingStrategy strategy) {
+        super(delegate);
+        this.strategy = requireNonNull(strategy, "strategy");
+    }
+
+    /**
+     * Invoked when {@code req} is not throttled. By default, this method delegates the specified {@code req} to
+     * the {@link #delegate()} of this service.
+     */
+    protected HttpResponse onSuccess(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        return delegate().serve(ctx, req);
+    }
+
+    /**
+     * Invoked when {@code req} is throttled. By default, this method responds with the
+     * {@link HttpStatus#SERVICE_UNAVAILABLE} status.
+     */
+    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req, @Nullable Throwable cause)
+            throws Exception {
+        final DefaultHttpResponse res = new DefaultHttpResponse();
+        res.respond(HttpStatus.UNAUTHORIZED);
```

Comment:
How about getting a throttled response status code from `ThrottlingStrategy`? A user might want to have their own throttling status code, so I think it would be better to be customizable.